### ISSUE TITLE
adapter-netlify missing dependency in output function

### ISSUE THREAD

**Issue by ixxie** [sveltejs/kit#1313]
Recently, I've been getting this weird error with `adapter-netlify`:
```
12:59:10 PM:   Error message
12:59:10 PM:   A Netlify Function is using "encoding" but that dependency has not been installed yet.
12:59:10 PM: ​
12:59:10 PM:   By default, dependencies inside a Netlify Function's "package.json" are not automatically installed.
  There are several ways to fix this problem:
12:59:10 PM:     - Removing your Function's "package.json" and adding the dependencies to the project's top-level "package.json" instead. This is the fastest and safest solution.
12:59:10 PM:     - Running "npm install" or "yarn" inside your Netlify Function in your build command.
12:59:10 PM:     - Adding the following plugin to your "netlify.toml":
12:59:10 PM: ​
12:59:10 PM:   [[plugins]]
12:59:10 PM:   package = "@netlify/plugin-functions-install-core"
```

Adding the suggested plugin doesn't seem to help...

I have not been able to reproduce the issue in a minimal repo, but my repo is public and [this commit](https://github.com/coflow-network/coflow-app/tree/04a9bffdd3feddf520cd553d85a6c3785dc032e6) reproduces the issue.

**Comment by Leftium** [comment]
I just tried deploying the latest SvelteKit demo to Netlify without issue: 
- https://sveltekit-demo-netlify-2021-05-04.netlify.app/
- https://github.com/Leftium/sveltekit-demo-netlify

So I don't think this is a SvelteKit issue, but rather an issue with your project.

---

Normally using `adapter-netlify@next` is recommended, but what you really want is the version that matches your SvelteKit template project. Because SvelteKit is in beta, there is no guarantee that the template project from Apr 3rd will be compatible with an adapter from Apr 30th, or even Apr 13th.

You may be able to fix the adapter issues in your project by installing a version of the adapter from around the date you initialized the project: https://www.npmjs.com/package/@sveltejs/adapter-netlify?activeTab=versions

However, what I would do is:
1. Start a new SvelteKit project with the latest template
2. Get the netlify adapter working
3. Apply your project's changes on top of this new project.

I have basically already done steps 1 and 2, so you can just clone/fork my repo:
https://github.com/Leftium/sveltekit-demo-netlify

**Comment by benmccann** [comment]
I can't run your demo. It just says `supabaseUrl is required`. But you should put `next` as the version for both SvelteKit and the adapter in your `package.json` and then delete your `package-lock.json` and run `npm install` again.

I'm going to close this since it's working in the default project and I can't reproduce

**Comment by Shaun-Regenbaum** [comment]
Im enountering this issue too.

**Comment by ixxie** [comment]
I tried setting SvelteKit and Adapter Netlify versions to `next` and reinstalling, and tried to study your fresh template repo to see if there is any difference in configuration that could cause this, but this all yielded nothing.

Can somebody explain to me what is responsible for building the `functions/` folder? It seems the root of the issue relates to `functions/render/index.js` is requiring a dependency that isn't specified anywhere. I get no import issues when running `npm run dev` and I am having trouble narrowing down the source of the issue...

**Comment by agarzola** [comment]
I just encountered this as well and have a workaround.

## TL;DR

Add `encoding` as a dependency on your project and your builds should pass:

```
// via npm
npm install --save encoding

// via yarn
yarn add encoding
```

## Some background

From the build logs in Netlify:
```
Error message
8:11:39 AM:   A Netlify Function is using "encoding" but that dependency has not been installed yet.
8:11:39 AM: ​
8:11:39 AM:   By default, dependencies inside a Netlify Function's "package.json" are not automatically installed.
  There are several ways to fix this problem:
8:11:39 AM:     - Removing your Function's "package.json" and adding the dependencies to the project's top-level "package.json" instead. This is the fastest and safest solution.
8:11:39 AM:     - Running "npm install" or "yarn" inside your Netlify Function in your build command.
8:11:39 AM:     - Adding the following plugin to your "netlify.toml":
```

I tried the topmost suggestion (adding `encoding` as a dependency of my project) and my build passed (hence my TL;DR solution above). Then I searched for the error message online and found [this support discussion](https://answers.netlify.com/t/error-could-not-find-encoding-module-in-file-netlify-function/2259/13) wherein a commenter says that `node-fetch` is the culprit, which led me down a rabbit hole…

Check if your project has a dependency (nested or otherwise) on `node-fetch`:

```
// via npm
npm ls node-fetch

// via yarn
yarn why node-fetch
```

In my project, I’m using `@prismicio/client`, which has a dependency on `node-fetch`. Here’s what `yarn why node-fetch` yields:

```
=> Found "node-fetch@2.6.1"
info Reasons this module exists
   - "@prismicio#client#cross-fetch" depends on it
   - Hoisted from "@prismicio#client#cross-fetch#node-fetch"
```

`node-fetch` [attempts to load `encoding`](https://github.com/node-fetch/node-fetch/blob/2.x/src/body.js#L14) (and catches the error if it cannot be found), but `encoding` is not listed as its dependency. I have zero experience with any of the tech involved in this (Svelte, SvelteKit, Vite, and Netlify Functions are all new to me! 😄), but I think that when the Netlify function is generated by the Netlify adapter, the `encoding` package is identified as a requirement and the generated file ends up with a hard dependency on it. Since the package is not in the list of dependencies for `node-fetch`, it is never installed.

This definitely is not technically an issue with SvelteKit or its Netlify adapter. It’s just the result of the way that functions are packaged and a failure of whatever shrink-wrapping function is being used for that to identify when a `require()` statement is composed in such a way as to make the dependency optional.

I hope this helps others running into this issue!

**Comment by ixxie** [comment]
Thank you @agarzola, this resolves the issue for me.

I wonder if an issue should be opened in the [node-fetch repo](https://github.com/node-fetch/node-fetch)?

I would be happy to do it, but you seem to have a better grasp of the topic.

**Comment by ixxie** [comment]
Right, I see your point. Well, I'll leave it to better minds to think about the issues of bundling and dependencies...

**Comment by samyoungnyc** [comment]
I'm experiencing the exact same problem but my error code is:

`A Netlify Function is using "exports" but that dependency has not been installed yet.`

Tried to install 'exports' but got the following error: 

`npm ERR! notsup Unsupported engine for super@0.1.0: wanted: {"node":">= 0.6.0 < 0.7.0"} (current: {"node":"14.17.0","npm":"6.14.13"})`

FWIW if anyone is interested I'd greatly appreciate help with this, I'm totally flummoxed.
